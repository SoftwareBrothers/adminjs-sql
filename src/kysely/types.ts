import { Kysely } from 'kysely';
import { ColumnMetadata, TableMetadata } from 'kysely-codegen';
import type { EnumCollection } from 'kysely-codegen/dist/core/enum-collection.js';

export type FindOptions = {
  limit?: number;
  offset?: number;
  sort?: {
    direction?: 'asc' | 'desc';
    sortBy?: string;
  };
};

export type Introspection = {
  enums: EnumCollection;
  tables: TableMetadata[];
}

export type DatabaseInformation = {
  _kind: symbol;
  connectionString: string;
  db: Kysely<any>;
  introspection: Introspection;
}

export type ResourceInformation = {
  _kind: symbol;
  db: Kysely<any>;
  introspection: Introspection;
  table: TableMetadata;
}

export type PropertyInformation = {
  _kind: symbol;
  column: ColumnMetadata;
  introspection: Introspection;
}

export const DATABASE_SYMBOL = Symbol('database');
export const RESOURCE_SYMBOL = Symbol('resource');
export const PROPERTY_SYMBOL = Symbol('property');
