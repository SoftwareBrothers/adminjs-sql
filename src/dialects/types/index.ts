import { Knex } from 'knex/types';

export type DatabaseDialect = 'postgresql' | 'mysql' | 'mysql2'

export interface BaseConnectionConfig {
    database: string;
    schema?: string;
}

export type ConnectionOptions =
    (Knex.PgConnectionConfig | Knex.MySqlConnectionConfig | Knex.MySql2ConnectionConfig)
    & BaseConnectionConfig;
