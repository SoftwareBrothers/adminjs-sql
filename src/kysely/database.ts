import { BaseDatabase, BaseResource } from 'adminjs';

import { Resource } from './resource.js';
import { DATABASE_SYMBOL, DatabaseInformation, RESOURCE_SYMBOL } from './types.js';

export class Database extends BaseDatabase {
  info: DatabaseInformation;

  constructor(info: DatabaseInformation) {
    super({});
    this.info = info;
  }

  static isAdapterFor(info: DatabaseInformation): boolean {
    return Boolean(
      info.db
      && info.db.constructor.name === 'Kysely'
      && info._kind === DATABASE_SYMBOL,
    );
  }

  resources(): Array<BaseResource> {
    return this.info.introspection.tables.map((table) => new Resource({
      _kind: RESOURCE_SYMBOL,
      introspection: this.info.introspection,
      db: this.info.db,
      table,
    }));
  }

  resource(name: string): BaseResource {
    const resources = this.resources();
    const resource = resources.find((resource) => resource.id() === name);
    if (!resource) {
      throw new Error(`Resource with id ${name} not found`);
    }

    return resource;
  }
}
