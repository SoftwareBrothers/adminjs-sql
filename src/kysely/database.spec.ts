import { describe, expect, it } from 'vitest';
import { BaseResource } from 'adminjs';
import { EnumCollection } from 'kysely-codegen/dist/core/enum-collection.js';

import { testFixtures } from '../test/fixtures.js';

import { Database } from './database.js';
import { DATABASE_SYMBOL } from './types.js';

const fixtures = testFixtures();

describe('database', () => {
  it('should return true if adapter is correct', () => {
    expect(
      Database.isAdapterFor({
        _kind: DATABASE_SYMBOL,
        connectionString: '',
        db: fixtures.db,
        introspection: {
          enums: new EnumCollection(),
          tables: [],
        },
      }),
    )
      .toBe(true);
  });

  it('should return resource', () => {
    expect(fixtures.database.resource('user'))
      .toBeInstanceOf(BaseResource);
  });

  it('should error when no table is found', async () => {
    expect(() => fixtures.database.resource('random'))
      .toThrowError('Resource with id random not found');
  });

  it('should return resources', () => {
    expect(fixtures.database.resources())
      .toMatchSnapshot();
  });
});
