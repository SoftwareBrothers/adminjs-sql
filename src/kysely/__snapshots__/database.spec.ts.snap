// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`database > should return resources 1`] = `
[
  Resource {
    "_decorated": null,
    "info": {
      "_kind": Symbol(resource),
      "db": Kysely {},
      "introspection": {
        "enums": EnumCollection {
          "enums": {
            "public.status": [
              "ACTIVE",
              "REMOVED",
            ],
          },
        },
        "tables": [
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": true,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "timestamp",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "created_at",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "title",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "content",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "timestamp",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "updated_at",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "json",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "some_json",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "status",
                "dataTypeSchema": "public",
                "enumValues": [
                  "ACTIVE",
                  "REMOVED",
                ],
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "status",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "author_id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "bool",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "published",
              },
            ],
            "name": "post",
            "schema": "public",
          },
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "uuid",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "bio",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "name",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "user_id",
              },
            ],
            "name": "profile",
            "schema": "public",
          },
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": true,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "email",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "name",
              },
            ],
            "name": "user",
            "schema": "public",
          },
        ],
      },
      "table": TableMetadata {
        "columns": [
          ColumnMetadata {
            "comment": null,
            "dataType": "int4",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": true,
            "isArray": false,
            "isAutoIncrementing": true,
            "isNullable": false,
            "name": "id",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "timestamp",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": true,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "created_at",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "text",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "title",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "text",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": true,
            "name": "content",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "timestamp",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": true,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "updated_at",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "json",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": true,
            "name": "some_json",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "status",
            "dataTypeSchema": "public",
            "enumValues": [
              "ACTIVE",
              "REMOVED",
            ],
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "status",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "int4",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "author_id",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "bool",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": true,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "published",
          },
        ],
        "name": "post",
        "schema": "public",
      },
    },
  },
  Resource {
    "_decorated": null,
    "info": {
      "_kind": Symbol(resource),
      "db": Kysely {},
      "introspection": {
        "enums": EnumCollection {
          "enums": {
            "public.status": [
              "ACTIVE",
              "REMOVED",
            ],
          },
        },
        "tables": [
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": true,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "timestamp",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "created_at",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "title",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "content",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "timestamp",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "updated_at",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "json",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "some_json",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "status",
                "dataTypeSchema": "public",
                "enumValues": [
                  "ACTIVE",
                  "REMOVED",
                ],
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "status",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "author_id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "bool",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "published",
              },
            ],
            "name": "post",
            "schema": "public",
          },
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "uuid",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "bio",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "name",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "user_id",
              },
            ],
            "name": "profile",
            "schema": "public",
          },
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": true,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "email",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "name",
              },
            ],
            "name": "user",
            "schema": "public",
          },
        ],
      },
      "table": TableMetadata {
        "columns": [
          ColumnMetadata {
            "comment": null,
            "dataType": "uuid",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": true,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "id",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "text",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "bio",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "text",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": true,
            "name": "name",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "int4",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "user_id",
          },
        ],
        "name": "profile",
        "schema": "public",
      },
    },
  },
  Resource {
    "_decorated": null,
    "info": {
      "_kind": Symbol(resource),
      "db": Kysely {},
      "introspection": {
        "enums": EnumCollection {
          "enums": {
            "public.status": [
              "ACTIVE",
              "REMOVED",
            ],
          },
        },
        "tables": [
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": true,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "timestamp",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "created_at",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "title",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "content",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "timestamp",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "updated_at",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "json",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "some_json",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "status",
                "dataTypeSchema": "public",
                "enumValues": [
                  "ACTIVE",
                  "REMOVED",
                ],
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "status",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "author_id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "bool",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "published",
              },
            ],
            "name": "post",
            "schema": "public",
          },
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "uuid",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "bio",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "name",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "user_id",
              },
            ],
            "name": "profile",
            "schema": "public",
          },
          TableMetadata {
            "columns": [
              ColumnMetadata {
                "comment": null,
                "dataType": "int4",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": true,
                "isArray": false,
                "isAutoIncrementing": true,
                "isNullable": false,
                "name": "id",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": false,
                "name": "email",
              },
              ColumnMetadata {
                "comment": null,
                "dataType": "text",
                "dataTypeSchema": "pg_catalog",
                "enumValues": null,
                "hasDefaultValue": false,
                "isArray": false,
                "isAutoIncrementing": false,
                "isNullable": true,
                "name": "name",
              },
            ],
            "name": "user",
            "schema": "public",
          },
        ],
      },
      "table": TableMetadata {
        "columns": [
          ColumnMetadata {
            "comment": null,
            "dataType": "int4",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": true,
            "isArray": false,
            "isAutoIncrementing": true,
            "isNullable": false,
            "name": "id",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "text",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": false,
            "name": "email",
          },
          ColumnMetadata {
            "comment": null,
            "dataType": "text",
            "dataTypeSchema": "pg_catalog",
            "enumValues": null,
            "hasDefaultValue": false,
            "isArray": false,
            "isAutoIncrementing": false,
            "isNullable": true,
            "name": "name",
          },
        ],
        "name": "user",
        "schema": "public",
      },
    },
  },
]
`;
